File: challenges/rate-limiter/typescript/solution/exercise.ts
import type { Clock } from "../../../../tooling/typescript/contracts.js";

export class SlidingWindowLimiter {
  // A head index avoids O(n) Array.shift on every expiration.
  private clients = new Map<string, { timestamps: number[]; head: number }>();

  constructor(private limit: number, private windowMs: number, private clock: Clock) {}

  allow(client: string): boolean {
    const now = this.clock();
    let history = this.clients.get(client);
    if (history === undefined) {
      history = { timestamps: [], head: 0 };
      this.clients.set(client, history);
    }
    while (history.head < history.timestamps.length &&
           history.timestamps[history.head] <= now - this.windowMs) {
      history.head++;
    }
    // Compact geometrically so storage remains O(limit) and cost is amortized.
    if (history.head > 0 && history.head * 2 >= history.timestamps.length) {
      history.timestamps = history.timestamps.slice(history.head);
      history.head = 0;
    }
    if (history.timestamps.length - history.head >= this.limit) return false;
    history.timestamps.push(now);
    return true;
  }
}

File: challenges/rate-limiter/go/solution/exercise.go
package exercise

import "refresher/tooling/go/contracts"

type SlidingWindowLimiter struct {
	limit    int
	windowMS int64
	clock    contracts.Clock
	clients  map[string][]int64
}

func NewSlidingWindowLimiter(limit int, windowMS int64, clock contracts.Clock) *SlidingWindowLimiter {
	return &SlidingWindowLimiter{limit, windowMS, clock, make(map[string][]int64)}
}

func (l *SlidingWindowLimiter) Allow(client string) bool {
	now := l.clock()
	history := l.clients[client]
	head := 0
	for head < len(history) && history[head] <= now-l.windowMS {
		head++
	}
	history = history[head:]
	if len(history) >= l.limit {
		l.clients[client] = history
		return false
	}
	l.clients[client] = append(history, now)
	return true
}
